src/core/ir/discriminator.rs

```diff
@@ -0,0 +1,1113 @@
+                        // if multiple fields are required in the same set of types than we can leave only one of such fields
+                        || (*field_info.required_in != 0 && seen_required_in.contains(&field_info.required_in));
+
+                    seen_required_in.insert(*field_info.required_in);
+
+                    !drop
+                })
+                .collect()
+        };
+
+        let discriminator = Self { fields_info, types };
+
+        tracing::debug!(
+            "Generated discriminator for union type '{union_name}':\n{discriminator:?}",
+        );
+
+        Valid::succeed(discriminator)
+    }
+
+    pub fn resolve_type(&self, value: &Value) -> Result<TypeName> {
+        if let Value::List(list) = value {
+            let results: Result<Vec<_>> = list
+                .iter()
+                .map(|item| Ok(self.resolve_type_for_single(item)?.to_string()))
+                .collect();
+
+            Ok(TypeName::Vec(results?))
+        } else {
+            Ok(TypeName::Single(
+                self.resolve_type_for_single(value)?.to_string(),
+            ))
+        }
+    }
+
+    fn resolve_type_for_single(&self, value: &Value) -> Result<&str> {
+        let Value::Object(obj) = value else {
+            bail!("Value expected to be object");
+        };
+
+        let mut possible_types = Repr::all_covered(self.types.len());
+
+        for (field, info) in &self.fields_info {
+            if obj.contains_key(field.as_str()) {
+                *possible_types &= *info.presented_in;
+            } else {
+                *possible_types &= !*info.required_in;
+            }
+
+            if possible_types.is_empty() {
+                // no possible types, something is wrong with the resolved value
+                bail!("Failed to find corresponding type for value")
+            }
+
+            if !possible_types.is_covering_multiple_types() {
+                // we've got only one possible type so return it
+                // even despite the fact the value could be totally wrong if we check other
+                // fields. But we want to cover positive cases and we want to do
+                // it as soon as possible and the wrong value will be probably
+                // wrong anyway to use later
+                return Ok(possible_types.first_covered_type(&self.types));
+            }
+        }
+
+        // we have multiple possible types. Return the first one
+        // that is defined earlier in config
+        Ok(possible_types.first_covered_type(&self.types))
+    }
+}
+
+/// Representation for set of types if some condition is hold.
+/// The condition is represented as a bit inside the usize number
+/// where bit position from the right in binary representation of usize
+/// is the index of type in the set and the if the value of the bit is
+/// 1 then the condition is hold
+#[derive(Copy, Clone, Default)]
+struct Repr(usize);
```

Comment:
Why not use a u128?

Reply:
u128 would imply using multiple registries and multiple cpu instructions to perform arithmetic operations while for usize I believe it is covered with single cpu registry and instruction